### PR TITLE
#160451459 create article functionality

### PR DESCRIPTION
**Pull request by onwuvic**
#### What does this PR do?
This pull request creates functionality to create an article in the frontend

#### Description of Task to be completed?
- Create JSX view form to create the article
- Create an action to create an article
- Create a reducer to create an article
- Write a test for the create article component

#### How should this be manually tested?
- Start the app
- Navigate to /api/v1/new-story
- create your article

#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
#160451459

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -41,14 +41,18 @@
...
     "jwt-decode": "^2.2.0",
     "materialize-css": "^1.0.0-rc.2",
+    "medium-editor": "^5.23.3",
     "npm-watch": "^0.4.0",
     "prop-types": "^15.6.2",
     "query-string": "^6.2.0",
     "react": "^16.6.0",
     "react-dom": "^16.6.0",
+    "react-idle-timer": "^4.0.9",
     "react-materialize": "^2.4.8",
```
Comment (onwuvic) [inline package.json:57]: medium-editor for text-area tag.
react-idle-timer for checking the idle time of the user before calling the auto-save function
react-tag-input for adding tags

File: src/components/Pages/LoginPage.js [hunk 2]
```diff
@@ -10,7 +10,7 @@ const LoginForm = () => (
       <label>Password: </label>
       <input type='password' />
 
-      <input type='submit' name='Login'/>
+      <input type='submit' name='Login' />
```
Comment (onwuvic) [inline src/components/Pages/LoginPage.js:13]: The position changed by eslint

File: src/constants/Constants.js [hunk 3]
```diff
@@ -1,6 +1,13 @@
 const BASE_URL = 'http://localhost:3000';
 export const articleConstant = {
-  ALL_ARTICLES: `${BASE_URL}/api/v1/articles/feed`
```
Comment (onwuvic) [inline src/constants/Constants.js:3]: The position changed by eslint

File: src/containers/createArticle/CreateArticle.js [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+      alertVisible: false,
+      isCreated: false,
+      tags: []
+    };
+
+    this.onInputChange = this.onInputChange.bind(this);
+    this.getAlertMessage = this.getAlertMessage.bind(this);
+    this.handleEditorChange = this.handleEditorChange.bind(this);
+    this.save = this.save.bind(this);
+    this.update = this.update.bind(this);
```
Comment (onwuvic) [inline src/containers/createArticle/CreateArticle.js:57]: I used bind because it won't work without binding

File: src/reducers/Index.js [hunk 5]
```diff
@@ -1,13 +1,20 @@
...
 import msgInfoReducer from './MsgInfoReducer';
+import categoryReducer from './CategoryReducer';
+import TagReducer from './TagReducer';
 
 
 const rootReducer = combineReducers({
   demoArticleReducer,
+  Article: articleReducer,
   User: userReducer,
-  Info: msgInfoReducer
```
Comment (onwuvic) [inline src/reducers/Index.js]: The position changed by eslint

File: src/styles/style.scss [hunk 6]
```diff
@@ -1,5 +1 @@
-$themeColor: #06576D;
-
-.theme-color {
-    color: $themeColor;
```
Comment (onwuvic) [inline src/styles/style.scss:4]: This style was not needed anymore

File: src/components/Pages/HomePage.js [hunk 7]
```diff
@@ -1,9 +1,15 @@
 import React from 'react';
+
 import '../../styles/style.scss';
+import Header from '../../containers/Header/Header';
+import Login from '../../containers/Login/Login';
+import MsgInfo from '../../containers/MsgInfo/MsgInfo';
 
 const HomePage = () => (
   <div>
-    <h1 className="theme-color">Hello! Welcome to Authors Haven</h1>
```
Comment (onwuvic) [inline src/components/Pages/HomePage.js]: This is not needed anymore

File: src/App.js [hunk 8]
```diff
@@ -1,44 +1,27 @@
...
    * @memberof App
    */
   render() {
     return (
       <div>
-        <Header />
-        <h1 className="theme-color">Hello! Welcome to Authors Haven</h1>
-        This is the app component
-        <Login />
-        <MsgInfo />
```
Comment (onwuvic) [inline src/App.js]: This is not needed anymore

File: src/App.js [hunk 9]
```diff
@@ -1,44 +1,27 @@
 import React, { Component } from 'react';
-import { Route } from 'react-router-dom';
-import Header from './containers/Header/Header';
-import Login from './containers/Login/Login';
-import Home from './containers/Home/Home';
```
Comment (onwuvic) [inline src/App.js]: These import are no longer needed in this file

File: src/App.js [hunk 10]
```diff
@@ -1,44 +1,27 @@
...
-import Navbar from './components/Navbar/Navbar';
-import LoginForm from './components/Pages/LoginPage';
-import HomePage from './components/Pages/HomePage';
-import MsgInfo from './containers/MsgInfo/MsgInfo';
+import { Route, Switch } from 'react-router-dom';
+// import Home from './containers/Home/Home';
```
Comment (s-akinrele) [inline src/App.js]: remove commented code

File: src/containers/Header/Header.js [hunk 11]
```diff
@@ -196,19 +197,25 @@ class Header extends Component {
...
+                  && <div className="sd-wrapper">
+                    <ul>
+                      <li className="s-list">
+                        <Link to="/new-story">
+                          <i className="fas fa-plus"></i>
+                          &nbsp;New Article
+                        </Link>
+                      </li>
+                      <li
+                        onClick={this.logout.bind(this)}
```
Comment (s-akinrele) [inline src/containers/Header/Header.js]: bind in the constructor

File: src/containers/createArticle/CreateArticle.js [hunk 12]
```diff
@@ -0,0 +1,353 @@
...
+    * @memberof Article
+    */
+  create() {
+    const {
+      hasChanges, title, description, body, category, image
+    } = this.state;
+
+    const categoryId = Number(category);
+
+    const tags = TagObjectToString(this.state.tags);
```
Comment (s-akinrele) [inline src/containers/createArticle/CreateArticle.js]: just destructure as you are doing above

File: src/containers/createArticle/CreateArticle.js [hunk 13]
```diff
@@ -0,0 +1,353 @@
...
+  /**
+    * @description - This method delete the Tag input values
+    * @param {objecj}  i
+    * @returns {object} void
+    * @memberof Article
+    */
+  handleDelete(i) {
+    const { tags } = this.state;
+    this.setState({
+      tags: tags.filter((tag, index) => index !== i),
```
Comment (s-akinrele) [inline src/containers/createArticle/CreateArticle.js]: rename  `i`

File: src/containers/createArticle/CreateArticle.js [hunk 14]
```diff
@@ -0,0 +1,353 @@
...
+    *  @param {string} newPos
+    * @returns {object} void
+    * @memberof Article
+    */
+  handleDrag(tag, currPos, newPos) {
+    const tags = [...this.state.tags];
+    const newTags = tags.slice();
+
+    newTags.splice(currPos, 1);
+    newTags.splice(newPos, 0, tag);
```
Comment (s-akinrele) [inline src/containers/createArticle/CreateArticle.js]: use the full spelling for readability purpose

File: src/containers/createArticle/CreateArticle.js (hunk 12 above)
Comment (onwuvic) [reply src/containers/createArticle/CreateArticle.js]: fixed

File: src/components/ArticleForm/ArticleForm.scss [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+    margin: 0;
+    .col{
+      padding: 0;
+    }
+  }
+}
+
+.main-content {
+  padding: 26px;
+}
```
Comment (s-akinrele) [inline src/components/ArticleForm/ArticleForm.scss]: add new line

File: src/containers/Header/Header.js [hunk 16]
```diff
@@ -32,6 +33,12 @@ class Header extends Component {
       isAuth: false,
     };
     this.timeoutID = null;
+    this.onShowDropDown = this.onShowDropDown.bind(this);
+    this.toggleDropDown = this.toggleDropDown.bind(this);
+    this.onBlur = this.onBlur.bind(this);
+    this.onLoginClicked = this.onLoginClicked.bind(this);
+    this.toggleSettingsOptions = this.toggleSettingsOptions.bind(this);
+    this.logout = this.logout.bind(this);
```
Comment (s-akinrele) [inline src/containers/Header/Header.js]: change this as discussed

File: src/containers/createArticle/CreateArticle.js (hunk 4 above)
Comment (s-akinrele) [reply src/containers/createArticle/CreateArticle.js:57]: arrange in alphabetical order

File: src/containers/createArticle/CreateArticle.js [hunk 17]
```diff
@@ -0,0 +1,298 @@
...
+
+  /**
+    * @description - This method delete the Tag input values
+    * @param {objecj}  tagToDelete
+    * @returns {object} void
+    * @memberof Article
+    */
+  handleDelete(tagToDelete) {
+    const { tags } = this.state;
+    this.setState({
```
Comment (s-akinrele) [inline src/containers/createArticle/CreateArticle.js:192]: tagIndexToDelete

File: src/containers/createArticle/CreateArticle.js [hunk 18]
```diff
@@ -0,0 +1,298 @@
...
+    newTags.splice(newPosition, 0, tag);
+    this.setState({ tags: newTags });
+  }
+
+  /**
+   * @description - This method renders the jsx for this component
+   * @returns {jsx} - jsx
+   * @memberof CreateArticle
+   */
+  render() {
```
Comment (s-akinrele) [inline src/containers/createArticle/CreateArticle.js:243]: const defaultValue = this.state.category ? this.state.category : 0

File: src/containers/createArticle/CreateArticle.js [hunk 19]
```diff
@@ -0,0 +1,298 @@
...
+                tags={this.state.tags}
+                suggestions={convertIdToString(this.props.suggestedTags)}
+                handleDelete={this.handleDelete}
+                handleAddition={this.handleAddition}
+                handleDrag={this.handleDrag}
+                required
+              />
+              <div className="input-field col s12">
+                <span htmlFor="Title">Categories</span>
+                <Input s={12} type='select' id="category" onChange={this.onInputChange} defaultValue={this.state.category ? this.state.category : '0'} required>
```
Comment (s-akinrele) [inline src/containers/createArticle/CreateArticle.js]:
```
<Input
 s={12}
 type='select'
 id="category"
 onChange={this.onInputChange}
 defaultValue={defaultValue}
 required
>
```

File: src/containers/createArticle/CreateArticle.scss [hunk 20]
```diff
@@ -0,0 +1,231 @@
...
+  div.pub-date {
+    font-size: 12px;
+    margin-top: 4px;
+    color: #777;
+  }
+
+  .publish-btn-wrapper{
+    text-align: right;
+  }
+
```
Comment (s-akinrele) [inline src/containers/createArticle/CreateArticle.scss:164]: remove unnecessary space

File: src/containers/createArticle/CreateArticle.scss [hunk 21]
```diff
@@ -0,0 +1,231 @@
...
+  }
+  .input-field .input-field {
+    margin-top: 9px;
+    background: #ffffff;
+    padding: 10px 10px 0;
+  }
+
+}
+
+
```
Comment (s-akinrele) [inline src/containers/createArticle/CreateArticle.scss]: remove this space

File: src/containers/createArticle/CreateArticle.test.js [hunk 22]
```diff
@@ -0,0 +1,45 @@
...
+const store = mockStore({
+  Article: {
+    article,
+    message: 'article created successfully',
+    tags,
+    loading: false,
+    error: true
+  },
+  Category: {
+    categories,
```
Comment (s-akinrele) [inline src/containers/createArticle/CreateArticle.test.js]: remove comma

**Review by s-akinrele — CHANGES_REQUESTED** [review]
I had some comments

File: src/containers/createArticle/CreateArticle.js (hunk 4 above)
Comment (onwuvic) [reply src/containers/createArticle/CreateArticle.js:57]: fixed

File: src/containers/createArticle/CreateArticle.js (hunk 17 above)
Comment (onwuvic) [reply src/containers/createArticle/CreateArticle.js:192]: fixed

File: src/containers/createArticle/CreateArticle.js (hunk 18 above)
Comment (onwuvic) [reply src/containers/createArticle/CreateArticle.js:243]: done

File: src/containers/createArticle/CreateArticle.scss (hunk 20 above)
Comment (onwuvic) [reply src/containers/createArticle/CreateArticle.scss:164]: fixed

File: src/containers/createArticle/CreateArticle.test.js (hunk 22 above)
Comment (onwuvic) [reply src/containers/createArticle/CreateArticle.test.js]: fixed

File: src/containers/createArticle/CreateArticle.scss (hunk 21 above)
Comment (onwuvic) [reply src/containers/createArticle/CreateArticle.scss]: fixed

File: src/containers/createArticle/CreateArticle.js (hunk 19 above)
Comment (onwuvic) [reply src/containers/createArticle/CreateArticle.js]: fixed

File: src/containers/createArticle/CreateArticle.js [hunk 23]
```diff
@@ -0,0 +1,332 @@
...
+
+    const categoryId = Number(category);
+
+    const tags = TagObjectToString(articleTags);
+
+    if (hasChanges) {
+      this.setState({ hasChanges: false, alertVisible: true, isCreated: true });
+
+      const formData = new FormData();
+      formData.append('title', title);
```
Comment (s-akinrele) [inline src/containers/createArticle/CreateArticle.js]: extract this into a method

File: src/containers/createArticle/CreateArticle.js [hunk 24]
```diff
@@ -0,0 +1,332 @@
...
+          startOnMount={false}
+          onIdle={this.save}>
+          <Row className="create-article">
+            <ArticleForm
+              componentState={this.state}
+              getAlertMessage={this.getAlertMessage()}
+              onPublishArticle={this.onPublishArticle}
+              onInputChange={this.onInputChange}
+              handleEditorChange={this.handleEditorChange}
+              onImageChange={this.onImageChange}
```
Comment (s-akinrele) [inline src/containers/createArticle/CreateArticle.js]: remove all these `on`

File: src/App.js (hunk 10 above)
Comment (onwuvic) [reply src/App.js]: fixed

**Review by s-akinrele — APPROVED** [review]

**Pull request merged**